refactor(ast): Remove `AstKind` for `MemberExpression` and replace with `StaticMemberExpression` and `PrivateFieldExpression`

camchenry opened the pull request:
- related to #11490 

Replaces the current `AstKind::MemberExpression` with `StaticMemberExpresion` and `PrivateFieldExpression`.

To make this change a little bit more ergonomic I've a new `MemberExpressionKind` which represents this subset of kinds.

camchenry commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11767?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11767** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11767?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11767?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11766** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11766?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11767).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 reviewed (commented):
should we introduce a `MemberExpression` enum so that we can go from `AstKind` to `MemberExpression`? it's such a common pattern in the ;linter

camchenry commented:
@camc314 @overlookmotel Thanks for the suggestion. I agree having the separate enum is a good idea, it makes the code a lot simpler and probably more maintainable in the future if ECMAScript should ever add more member expression types.

I've gone ahead and manually implemented an API for this, but an autogenerated one in the future would be nice.

overlookmotel commented:
I've hopefully fixed `no-console`, but I'm out of time, and am away for the next week. @camc314 can I please hand over to you to review?

camchenry committed:
refactor(ast): Remove `AstKind` for `MemberExpression` and replace with `StaticMemberExpression` and `PrivateFieldExpression` (#11767)

- related to #11490

Replaces the current `AstKind::MemberExpression` with `StaticMemberExpresion` and `PrivateFieldExpression`.

To make this change a little bit more ergonomic I've a new `MemberExpressionKind` which represents this subset of kinds.
